### DISCUSSION TITLE
Type checking and compilation

### DISCUSSION THREAD

**Discussion by travispaxton** [xtclang/xvm#253] [General]
I’ve been reading through your docs and looking through the source code. How are you handing type checking? Are you deferring to Java? I see you have a list of op codes and it looks like there’s some Java code that is emitting them, but I’m not quite grasping how you’re compiling.

**Comment by cpurdy** [comment]
Hi Travis -

No, we're not deferring to Java for type checking. When we compile, we produce a record of each type, called a [TypeConstant](https://github.com/xtclang/xvm/blob/master/javatools/src/main/java/org/xvm/asm/constants/TypeConstant.java), and that is used for type checking. It's a huge pile of information (including a tree of sub-classes of TypeConstant, and a separate structure produced from TypeConstant called [TypeInfo](https://github.com/xtclang/xvm/blob/master/javatools/src/main/java/org/xvm/asm/constants/TypeInfo.java)).

The type checking looks like [this in IfStatement](https://github.com/xtclang/xvm/blob/45b7dbc4e3c0efd4787589dff4a842a8925eec1d/javatools/src/main/java/org/xvm/compiler/ast/IfStatement.java#L162), where it knows it needs a condition to be a boolean, for example. I gave that as a super simple example, because most type checking is dramatically more complex than that, but it's mostly implemented in the validation stage of the compiler, by the various AST nodes.
